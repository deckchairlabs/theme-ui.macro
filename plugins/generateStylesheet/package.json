{
  "main": "dist/theme-ui.macro-plugins-generateStylesheet.cjs.js"
}
