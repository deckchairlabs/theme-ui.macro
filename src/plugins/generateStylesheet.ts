import * as fs from 'fs'
import * as Babel from '@babel/core'
import { MacroError } from 'babel-plugin-macros'
import { Theme, css, CSSObject } from '@theme-ui/css'
import * as PostCSS from 'postcss'
import nested from 'postcss-nested'
import {
  camelCaseToKebabCase,
  notUndefined,
  primitiveToCssValue,
} from '../utils'
import { Plugin, GenerateStylesheetPluginConfig } from '../types'
import { isStringLiteral } from '@babel/types'

export default function GenerateStylesheetPlugin(
  config: GenerateStylesheetPluginConfig
): Plugin {
  const { output, selectors } = config
  return (
    path: Babel.NodePath<Babel.Node>,
    theme: Theme,
    babel: typeof Babel
  ) => {
    const stylesheet = PostCSS.root()

    if (!babel.types.isObjectExpression(path.node)) {
      throw new MacroError('Expected an object expression')
    }

    const customProperties = resolveCustomPropertyDeclarations(path)

    if (customProperties.size > 0) {
      const root = PostCSS.rule({
        selector: ':root',
        nodes: Array.from(customProperties.values()),
      })

      stylesheet.append(root)
    }

    const nodes = resolveSelectorNodes(selectors, theme)

    stylesheet.append(nodes)

    const result = stylesheet.toResult({ to: output })
    const processor = PostCSS.default([nested()])
    const processed = processor.process(result)

    if (result.opts.to) {
      fs.writeFileSync(result.opts.to, processed.css)
    }
  }
}

function resolveCustomPropertyDeclarations(path: Babel.NodePath<Babel.Node>) {
  const customProperties: Map<string, PostCSS.Declaration> = new Map()

  function addCustomProperty(prop: string, value: string) {
    customProperties.set(
      prop,
      PostCSS.decl({
        prop,
        value,
      })
    )
  }

  path.traverse({
    ObjectProperty: {
      enter(nodePath) {
        if (
          isStringLiteral(nodePath.node.value) &&
          isStringLiteral(nodePath.node.key) &&
          nodePath.node.key.value.startsWith('--')
        ) {
          addCustomProperty(nodePath.node.key.value, nodePath.node.value.value)
        }
      },
    },
  })

  return customProperties
}

function resolveSelectorNodes(selectors: Record<string, string>, theme: Theme) {
  return Object.entries(selectors)
    .map(([key, selector]) => {
      const object = theme[key as keyof Theme]
      if (typeof object === 'object') {
        return objectToPostCSSNode(theme, object, selector)
      }
    })
    .filter(notUndefined)
}

function objectToPostCSSNode(theme: Theme, object: object, selector: string) {
  const nodes: PostCSS.Node[] = Object.entries(object)
    .map(([key, value]) => {
      if (Array.isArray(value)) {
        const styles = css({ [key]: value })(theme)
        return cssObjectToPostCSSDeclarations(styles)
      } else {
        switch (typeof value) {
          case 'string':
          case 'number':
            const styles = css({ [key]: value })(theme)
            return cssObjectToPostCSSDeclarations(styles)
          case 'object':
            const isPseudo = key.startsWith(':')
            const separator = isPseudo ? '' : '-'
            return objectToPostCSSNode(theme, value, `&${separator}${key}`)
        }
      }
    })
    .filter(notUndefined)
    .flat()

  return PostCSS.rule({
    selector,
    nodes,
  })
}

function cssObjectToPostCSSDeclarations(css: CSSObject) {
  const declarations: PostCSS.Node[] = Object.entries(css)
    .map(([key, value]) => {
      if (key.startsWith('@media')) {
        return PostCSS.atRule({
          name: 'media',
          params: key.replace('@media', '').trim(),
          nodes: cssObjectToPostCSSDeclarations(value as CSSObject),
        })
      } else {
        const prop = camelCaseToKebabCase(key)
        const cssValue = primitiveToCssValue(value as string | number)
        return PostCSS.decl({
          prop,
          value: cssValue,
        })
      }
    })
    .filter(notUndefined)

  return declarations
}
