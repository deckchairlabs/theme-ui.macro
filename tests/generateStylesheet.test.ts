import path from 'path'
import pluginTester from 'babel-plugin-tester'
import plugin from 'babel-plugin-macros'
import generateStylesheet from '../src/plugins/generateStylesheet'

pluginTester({
  plugin,
  pluginName: 'theme-ui.macro',
  snapshot: false,
  babelOptions: {
    presets: ['@babel/preset-typescript'],
    filename: __filename,
  },
  pluginOptions: {
    themeUI: {
      plugins: [
        generateStylesheet({
          selectors: { buttons: '.button', layout: '.layout' },
          output: path.resolve(__dirname, 'generated/theme.css'),
        }),
      ],
    },
  },
  fixtures: path.join(__dirname, 'plugins/generateStylesheet'),
})
