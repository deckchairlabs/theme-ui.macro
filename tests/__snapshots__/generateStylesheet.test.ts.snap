// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`generated stylesheet matches snapshot 1`] = `
:root {
    --colors-primary: red
}
.button-base {
    padding: 16px;
    padding-left: 8px;
    padding-right: 8px;
}
@media screen and (min-width: 40em) {
    .button-base {
        padding-left: 16px;
        padding-right: 16px
    }
}
.button-primary {
    background-color: red
}
.layout-spacing-large {
    margin: 16px
}
`;
